mirgecom/euler.py

```diff
@@ -20,20 +22,335 @@
+# )
 
 __doc__ = """
-.. autofunction:: euler_operator
+.. autofunction:: inviscid_operator
+.. autofunction:: number_of_scalars
+.. autofunction:: split_conserved
+.. autofunction:: split_species
+.. autofunction:: split_fields
+.. autofunction:: get_inviscid_timestep
 """
 
 
-def euler_operator(discr, w):
-    pass
+#
+# Euler flow eqns:
+# d_t(q) + nabla .dot. f = 0 (no sources atm)
+# state vector q: [rho rhoE rhoV]
+# flux tensor f: [rhoV (rhoE + p)V (rhoV.x.V + p*I)]
+#
+
+
+@dataclass
+class ConservedVars:
+    r"""
+    Class to resolve the canonical conserved quantities,
+    (mass, energy, momentum) per unit volume =
+    :math:`(\rho,\rhoE,\rho\vec{V})` from an agglomerated
+    object array.
+
+    .. attribute:: mass
+
+        Mass per unit volume
+
+    .. attribute:: energy
+
+        Energy per unit volume
+
+    .. attribute:: momentum
+
+        Momentum vector per unit volume
+    """
+    mass: np.ndarray
+    energy: np.ndarray
+    momentum: np.ndarray
+
+
+@dataclass
+class MassFractions:
+    r"""
+    Class to pick off the species mass fractions
+    (mass fractions) per unit volume =
+    :math:`(\rhoY_{\alpha}) | 1 \le \alpha \le N_{species}`,
+    from an agglomerated object array. :math:`N_{species}` is
+    the number of mixture species.
+
+    .. attribute:: mass
+
+        Mass fraction per unit volume for each mixture species
+    """
+    massfractions: np.ndarray
+
+
+def split_fields(ndim, q):
+    """
+    Method to spit out a list of named flow variables in
+    an agglomerated flow solution. Useful for specifying
+    named data arrays to helper functions (e.g. I/O).
+    """
+    qs = split_conserved(ndim, q)
+    mass = qs.mass
+    energy = qs.energy
+    mom = qs.momentum
+
+    retlist = [
+        ("mass", mass),
+        ("energy", energy),
+        ("momentum", mom),
+    ]
+    nscalar = number_of_scalars(ndim, q)
+    if nscalar > 0:
+        massfrac = split_species(ndim, q).massfraction
+        retlist.append(("massfraction", massfrac))
+
+    return retlist
+
+
+def number_of_scalars(ndim, q):
+    """
+    Return the number of scalars or mixture species in a flow solution.
+    """
+    return len(q) - (ndim + 2)
+
+
+def number_of_equations(ndim, q):
+    """
+    Return the number of equations (i.e. number of dofs) in the soln
+    """
+    return len(q) + number_of_scalars(ndim, q)
+
+
+def split_conserved(dim, q):
+    """
+    Return a 'ConservedVars' object that splits out conserved quantities
+    by name. Useful for expressive coding.
+    """
+    return ConservedVars(mass=q[0], energy=q[1], momentum=q[2:2+dim])
+
+
+def split_species(dim, q):
+    """
+    Return a 'MassFractions' object that splits out mixture species
+    conserved quantities by name. Useful for expressive coding.
+    """
+    numscalar = number_of_scalars(dim, q)
+    sindex = dim + 2
+    return MassFractions(massfractions=q[sindex:sindex+numscalar])
+
+
+def _inviscid_flux(discr, q, eos=IdealSingleGas()):
```

Comment:
I'm not sure I like that `eos` has a default. I don't think it's ever used without one. In addition, I'd like the argument order be such that the state vector is last.